segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,457 @@
+    }
+
+    /**
+     * Reports the metrics periodically for each EventProcessor.
+     */
+    private void reportMetrics() {
+        for (EventProcessorImpl ep : eventProcessorMap.values()) {
+            SegmentStoreMetrics.outstandingEventProcessorBytes(ep.getName(), containerId, ep.outstandingBytes.get());
+            // Only report the last iteration processing latency if it has been successful.
+            if (!ep.failedIteration.get()) {
+                ep.metrics.batchProcessingLatency(ep.lastIterationLatency.get());
+            }
+        }
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        if (!closed.getAndSet(true)) {
+            log.info("{}: Closing ContainerEventProcessor service.", this.traceObjectId);
+            closeProcessors();
+            eventProcessorMetricsReporting.cancel(true);
+        }
+    }
+
+    private void closeProcessors() {
+        synchronized (eventProcessorMap) {
+            if (eventProcessorMap.isEmpty()) {
+                return;
+            }
+            eventProcessorMap.forEach((k, v) -> {
+                try {
+                    v.close();
+                } catch (Exception e) {
+                    log.warn("{}: Problem closing EventProcessor {}.", this.traceObjectId, k, e);
+                }
+            });
+            eventProcessorMap.clear();
+            log.debug("{}: Closing EventProcessors complete.", this.traceObjectId);
+        }
+    }
+
+    //endregion
+
+    //region ContainerEventProcessor Implementation
+
+    @Override
+    public CompletableFuture<EventProcessor> forConsumer(@NonNull String name,
+                                                         @NonNull Function<List<BufferView>, CompletableFuture<Void>> handler,
+                                                         @NonNull EventProcessorConfig config) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+
+        synchronized (eventProcessorMap) {
+            // If the EventProcessor is already loaded, just return it.
+            if (eventProcessorMap.containsKey(name)) {
+                return CompletableFuture.completedFuture(eventProcessorMap.get(name));
+            }
+
+            // Instantiate the EventProcessor and put it into the map. If the EventProcessor is closed, auto-unregister.
+            Runnable onClose = () -> eventProcessorMap.remove(name);
+            return segmentSupplier.apply(name)
+                    .thenApply(segment -> {
+                        Exceptions.checkNotClosed(this.closed.get(), this);
+                        EventProcessorImpl eventProcessor = new EventProcessorImpl(name, segment, handler, config, onClose);
+                        eventProcessorMap.put(name, eventProcessor);
+                        eventProcessor.run(); // Start the processing for this EventProcessor and return it.
```

Comment:
There are 2 cases here:
- If the failure happens before the `Futures.loop()` starts working, the call itself to `forConsumer()` will throw an error to the caller.
- If the error happens once the process is running, like errors writing or reading the Segment, these will be handled by the `EventProcessorImpl` handle in `processEvents()`.
- 
In any case, thanks to your comment now the `run()` is execute _before_ putting the `EventProcessor` in the map (it was not the case before). This could have lead to having failed processors in the map (in case that  the `run()` method failed).